File: src/main/java/com/rdvdev2/timetravelmod/impl/ModTimeMachines.java
package com.rdvdev2.timetravelmod.impl;

import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachine;
import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachineUpgrade;
import com.rdvdev2.timetravelmod.api.timemachine.TimeMachineBuilder;
import com.rdvdev2.timetravelmod.impl.common.timemachine.upgrade.TimeMachineTrackerUpgrade;
import net.minecraft.block.BlockState;
import net.minecraft.util.registry.Registry;

public class ModTimeMachines {

    private static final BlockState[] BASIC_TIME_MACHINE_BLOCKS = new BlockState[]{
            ModBlocks.TIME_MACHINE_BASIC_BLOCK.getDefaultState(),
            ModBlocks.TIME_MACHINE_CABLE_PANEL.getDefaultState(),
            ModBlocks.TIME_MACHINE_PILLAR.getDefaultState(),
            ModBlocks.TIME_MACHINE_HORIZONTAL_VENTILATION.getDefaultState(),
            ModBlocks.TIME_MACHINE_VERTICAL_VENTILATION.getDefaultState()
    };

    public static final ITimeMachine CREATIVE = new TimeMachineBuilder()
            .setIcon(ModItems.CREATIVE_TIME_MACHINE)
            .setCooldownTime(0)
            .setStructureLayers(new String[][]{{"*"}})
            .setControllerStates()
            .setCoreStates()
            .setBasicStates()
            .build();

    public static final ITimeMachine TIER_1 = new TimeMachineBuilder()
            .setStructureLayers(new String[][]{
                    {
                            "BBB",
                            "BBB",
                            "BBB"
                    }, {
                            "BZB",
                            "B B",
                            "B B"
                    }, {
                            "BBB",
                            "B B",
                            "B B"
                    }, {
                            "BBB",
                            "BCB",
                            "BBB"
                    }})
            .setControllerStates(ModBlocks.TIME_MACHINE_CONTROL_PANEL.getDefaultState())
            .setCoreStates(ModBlocks.TIME_MACHINE_CORE.getDefaultState())
            .setBasicStates(BASIC_TIME_MACHINE_BLOCKS)
            .build();

    public static final ITimeMachine BIG_TIER_1 = new TimeMachineBuilder()
            .setStructureLayers(new String[][]{
                    {
                            "BBBBB",
                            "BBBBB",
                            "BBBBB",
                            "BBBBB",
                            "BBBBB"
                    }, {
                            "BBBBB",
                            "B   B",
                            "B   B",
                            "B   B",
                            "BBBBB"
                    }, {
                            "BBZBB",
                            "B   B",
                            "B   B",
                            "B   B",
                            "BB BB"
                    }, {
                            "BBBBB",
                            "B   B",
                            "B   B",
                            "B   B",
                            "BB BB"
                    }, {
                            "BBBBB",
                            "BBBBB",
                            "BBCBB",
                            "BBBBB",
                            "BBBBB"
                    }})
            .setControllerStates(ModBlocks.TIME_MACHINE_CONTROL_PANEL.getDefaultState())
            .setCoreStates(ModBlocks.TIME_MACHINE_CORE.getDefaultState())
            .setBasicStates(BASIC_TIME_MACHINE_BLOCKS)
            .setCorruptionMultiplier(3)
            .setEntityMaxLoad(3)
            .setCooldownTime(60 * 20)
            .build();

    public static void register() {
        registerTimeMachine("creative", CREATIVE);
        registerTimeMachine("tier_1", TIER_1);
        registerTimeMachine("big_tier_1", BIG_TIER_1);
        Upgrades.register();
    }

    public static void registerTimeMachine(String path, ITimeMachine timeMachine) {
        Registry.register(ModRegistries.TIME_MACHINE, Mod.identifier(path), timeMachine);
    }

    public static class Upgrades {

        public static final ITimeMachineUpgrade TRACKER = new TimeMachineTrackerUpgrade();

        public static void register() {
            registerTimeMachineUpgrade("tracker", TRACKER);
        }

        public static void registerTimeMachineUpgrade(String path, ITimeMachineUpgrade timeMachineUpgrade) {
            Registry.register(ModRegistries.TIME_MACHINE_UPGRADE, Mod.identifier(path), timeMachineUpgrade);
        }
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/impl/ModTimeMachines.kt
package com.rdvdev2.timetravelmod.impl

import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachine
import com.rdvdev2.timetravelmod.api.timemachine.ITimeMachineUpgrade
import com.rdvdev2.timetravelmod.api.timemachine.TimeMachineBuilder
import com.rdvdev2.timetravelmod.impl.Mod.identifier
import com.rdvdev2.timetravelmod.impl.common.timemachine.upgrade.TimeMachineTrackerUpgrade
import net.minecraft.util.registry.Registry

object ModTimeMachines {

    private val BASIC_TIME_MACHINE_BLOCKS = arrayOf(
        ModBlocks.TIME_MACHINE_BASIC_BLOCK.defaultState,
        ModBlocks.TIME_MACHINE_CABLE_PANEL.defaultState,
        ModBlocks.TIME_MACHINE_PILLAR.defaultState,
        ModBlocks.TIME_MACHINE_HORIZONTAL_VENTILATION.defaultState,
        ModBlocks.TIME_MACHINE_VERTICAL_VENTILATION.defaultState
    )

    val CREATIVE = TimeMachineBuilder()
        .setIcon(ModItems.CREATIVE_TIME_MACHINE)
        .setCooldownTime(0)
        .setStructureLayers(arrayOf(arrayOf("*")))
        .setControllerStates()
        .setCoreStates()
        .setBasicStates()
        .build()

    val TIER_1 = TimeMachineBuilder()
        .setStructureLayers(
            arrayOf(
                arrayOf(
                    "BBB",
                    "BBB",
                    "BBB"
                ), arrayOf(
                    "BZB",
                    "B B",
                    "B B"
                ), arrayOf(
                    "BBB",
                    "B B",
                    "B B"
                ), arrayOf(
                    "BBB",
                    "BCB",
                    "BBB"
                )
            )
        )
        .setControllerStates(ModBlocks.TIME_MACHINE_CONTROL_PANEL.defaultState)
        .setCoreStates(ModBlocks.TIME_MACHINE_CORE.defaultState)
        .setBasicStates(*BASIC_TIME_MACHINE_BLOCKS)
        .build()

    val BIG_TIER_1 = TimeMachineBuilder()
        .setStructureLayers(
            arrayOf(
                arrayOf(
                    "BBBBB",
                    "BBBBB",
                    "BBBBB",
                    "BBBBB",
                    "BBBBB"
                ), arrayOf(
                    "BBBBB",
                    "B   B",
                    "B   B",
                    "B   B",
                    "BBBBB"
                ), arrayOf(
                    "BBZBB",
                    "B   B",
                    "B   B",
                    "B   B",
                    "BB BB"
                ), arrayOf(
                    "BBBBB",
                    "B   B",
                    "B   B",
                    "B   B",
                    "BB BB"
                ), arrayOf(
                    "BBBBB",
                    "BBBBB",
                    "BBCBB",
                    "BBBBB",
                    "BBBBB"
                )
            )
        )
        .setControllerStates(ModBlocks.TIME_MACHINE_CONTROL_PANEL.defaultState)
        .setCoreStates(ModBlocks.TIME_MACHINE_CORE.defaultState)
        .setBasicStates(*BASIC_TIME_MACHINE_BLOCKS)
        .setCorruptionMultiplier(3)
        .setEntityMaxLoad(3)
        .setCooldownTime(60 * 20)
        .build()

    fun register() {
        CREATIVE.registerAs("creative")
        TIER_1.registerAs("tier_1")
        BIG_TIER_1.registerAs("big_tier_1")
        Upgrades.register()
    }

    object Upgrades {

        val TRACKER: ITimeMachineUpgrade = TimeMachineTrackerUpgrade()

        fun register() {
            TRACKER.registerAs("tracker")
        }

        private fun ITimeMachineUpgrade.registerAs(path: String) = Registry.register(ModRegistries.TIME_MACHINE_UPGRADE, identifier(path), this)
    }

    private fun ITimeMachine.registerAs(path: String) = Registry.register(ModRegistries.TIME_MACHINE, identifier(path), this)
}